google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreBundle.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.firestore;
+
+import com.google.api.client.util.Lists;
+import com.google.cloud.Timestamp;
+import com.google.firestore.proto.BundleElement;
+import com.google.firestore.proto.BundleMetadata;
+import com.google.firestore.proto.BundledDocumentMetadata;
+import com.google.firestore.proto.BundledQuery;
+import com.google.firestore.proto.NamedQuery;
+import com.google.firestore.v1.Document;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.util.JsonFormat;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/** Represents a Firestore data bundle with results from the given document and query snapshots. */
+public final class FirestoreBundle {
+
+  static int BUNDLE_SCHEMA_VERSION = 1;
+
+  // Raw byte array to hold the content of the bundle.
+  private byte[] bundleData;
+
+  /** Builds a Firestore data bundle with results from the given document and query snapshots. */
+  public static final class Builder {
+    // Id of the bundle.
+    private String id;
+    // Resulting documents for the bundle, keyed by full document path.
+    private Map<String, BundledDocument> documents = new HashMap<>();
+    // Named queries saved in the bundle, keyed by query name.
+    private Map<String, NamedQuery> namedQueries = new HashMap<>();
+    // The latest read time among all bundled documents and queries.
+    private Timestamp latestReadTime = Timestamp.MIN_VALUE;
+    // Printer to encode protobuf objects into JSON string.
+    private JsonFormat.Printer printer = JsonFormat.printer();
```

Comment:
If JsonFormat.Printer is thread safe (a quick glance though it seems like it is, but I don't know for sure), this field could be turned into a constant and have a single instance shared across all builders.